### PR TITLE
BOP-Elites

### PR DESCRIPTION
**Pull request by szhaovas**
## Description

Implement [BOP-Elites](https://arxiv.org/abs/2307.09326).

## TODO

- [x] Implement an emitter that maintains GP surrogrates on the objective and measure functions, and emits solutions maximizing the *Expected Joint Improvement of Elites* (EJIE) acquisition function.
- [x] Implement a scheduler to accommodate archive upscaling.
- [x] Write a simple script that demos how to run a BOP-Elites experiment.

## Questions

- [x] None

## Status

- [x] I have read the guidelines in
      [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md)
- [x] I have formatted my code using `yapf`
- [x] I have tested my code by running `pytest`
- [x] I have linted my code with `pylint`
- [x] I have added a one-line description of my change to the changelog in
      `HISTORY.md`
- [x] This PR is ready to go

### CODE REVIEW

File: boe_test.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    archive = GridArchive(
+        solution_dim=sphere_domain.solution_dim,
+        dims=params.upscale_schedule[0],
+        ranges=[[-5.12, 5.12]] * 2,
+        seed=params.seed,
+    )
+    # The passive archive that does NOT interact with BOP-Elites other than
+    # storing all evaluated data seen so far. We do not scale the passive
+    # archive, and it always stays at the final resolution.
+    passive_archive = GridArchive(
```
Comment (btjanaka) [inline boe_test.py:132]: Is it possible for this to be passed as the `result_archive` to the scheduler? That would be more consistent with how CMA-MAE currently operates.

File: boe_test.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        size=(params.init_nsamples, sphere_domain.solution_dim),
+    )
+    init_obj, init_meas = sphere_domain.evaluate(init_sol)
+
+    # The main component of BOP-Elites
+    emitters = [
+        BayesianOptimizationEmitter(
+            archive=archive,
+            init_data=(init_sol, init_obj, init_meas),
+            bounds=[(-10.24, 10.24)] * sphere_domain.solution_dim,
```
Comment (btjanaka) [inline boe_test.py:152]: Since this is a demo script, I think it's better to just put the params in here directly. Users can then figure out how to configure their parameters on their own later.

Comment (btjanaka) [reply boe_test.py:152]: The general idea of an example is to make something that's fairly clear and easy to understand.

File: boe_test.py
[the remark this answers is not in the document]
Comment (btjanaka) [reply boe_test.py:1]: Or maybe we can even turn it into a tutorial? What do you think of making this a Jupyter notebook? I can help write the tutorial after you edit the script. We can also reach out to Paul to get his approval.

File: boe_test.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    # archive, and it always stays at the final resolution.
+    passive_archive = GridArchive(
+        solution_dim=sphere_domain.solution_dim,
+        dims=params.upscale_schedule[-1],
+        ranges=[[-5.12, 5.12]] * 2,
+        seed=params.seed,
+    )
+
+    # An initial batch of data used to warm up GPs
+    init_sol = np.random.default_rng(params.seed).normal(
```
Comment (btjanaka) [inline boe_test.py:140]: Our `GaussianEmitter` outputs initial solutions from a Gaussian distribution on the first call to it since there are no solutions available in the archive for mutation. This mechanism is very similar to what is done here. I suggest taking a look at that since that would remove the extra solution generation and archive additions.

File: ribs/archives/_archive_base.py [hunk 4]
```diff
@@ -291,6 +294,23 @@ def clear(self):
         self._store.clear()
         self._stats_reset()
 
+    @abstractmethod
+    def retessellate(self, new_tessellation):
```
Comment (btjanaka) [inline ribs/archives/_archive_base.py:298]: I'm pretty hesitant about adding this method since:
1. It seems really vague -- `new_tessellation` can literally be anything.
2. This method is only really used in `GridArchive`
3. It assumes archives will be based on tessellations, which is not always true, e.g., see `ProximityArchive`.

Let's figure out an alternative once you implement the other changes. Perhaps we can just add this to certain archives like `GridArchive`.

File: ribs/archives/_grid_archive.py [hunk 5]
```diff
@@ -155,6 +158,55 @@ def boundaries(self):
...
+            solution_dim=self.solution_dim,
+            dims=new_tessellation,
+            ranges=list(zip(self.lower_bounds, self.upper_bounds)),
+            learning_rate=None,
+            threshold_min=-np.inf,
+            epsilon=self.epsilon,
+            qd_score_offset=self.qd_score_offset,
+            seed=self._seed,
+            dtype=self.dtypes,
+            # extra_fields=None,
```
Comment (btjanaka) [inline ribs/archives/_grid_archive.py:200]: `extra_fields` would need to be included. As stated on my `ArchiveBase` comment, I think there may be a better way to do this method, but let's come back to it later.

**Review by btjanaka — CHANGES_REQUESTED** [review]
Thanks @szhaovas! Left a bunch of comments. I haven't really looked at the emitter and scheduler themselves yet; I'll wait for you to fix the big things like how the scheduler works before taking a closer look.

File: setup.py [hunk 6]
```diff
@@ -35,23 +35,22 @@
         ### visualize ###
         "matplotlib>=3.0.0",
         "shapely>=2.0.0",
-
         ### cma ###
         "cma",
+        ### pymoo ###
```
[the remark this answers is not in the document]
Comment (szhaovas) [reply setup.py:40]: > do we actually use pymoo in BOP-Elites?

BOP-Elites uses [PatternSearch](https://pymoo.org/algorithms/soo/pattern.html) optimizer to optimize EJIE acquisition. I initially considered making our own pattern search implementation to avoid adding extra dependencies, but we might eventually need pymoo anyways if we decide to add multi-objective optimization to pyribs.

File: setup.py [hunk 7]
```diff
@@ -70,12 +69,13 @@
         "sphinxcontrib-jsmath==1.0.1",
         "sphinxcontrib-qthelp==1.0.3",
         "sphinxcontrib-serializinghtml==1.1.5",
-
         # Distribution
         "bump2version==1.0.1",
         "wheel==0.40.0",
         "twine==4.0.2",
         "check-wheel-contents==0.4.0",
+        # wandb
```
[the remark this answers is not in the document]
Comment (szhaovas) [reply setup.py:77]: Only for `boe_test.py`, I will rewrite `boe_test.py` into an example and remove `wandb` as a dependency.

File: ribs/archives/_archive_base.py (hunk 4 above)
Comment (szhaovas) [reply ribs/archives/_archive_base.py:298]: > Perhaps we can just add this to certain archives like GridArchive.

Yes, I think that might be best. I can remove `retessellate ` from `ArchiveBase` and have it only under `GridArchive`.

File: boe_test.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    name="BOP-Elites",
+    id=str(uuid.uuid5(uuid.NAMESPACE_DNS, str(asdict(params)))),
+    tags=["BOP-Elites", "Sphere"],
+    resume="allow",  # Allow resuming from same run ID
+)
+
+# Sphere domain evaluator
+sphere_domain = Sphere(solution_dim=4, shift=2)
+
+reload_checkpoint = None
```
[the remark this answers is not in the document]
Comment (szhaovas) [reply boe_test.py:118]: Yeah the logging of archive heatmaps and checkpoints are mostly for dev convenience. I can remove those as well as the reload in the re-written `boe_test`.

File: boe_test.py
[the remark this answers is not in the document]
Comment (szhaovas) [reply boe_test.py:1]: Let's have an example first.

File: ribs/schedulers/_bayesian_opt_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            if not isinstance(e, BayesianOptimizationEmitter):
+                raise TypeError(
+                    "All emitters must be of type BayesianOptimizationEmitter, "
+                    f"but emitter{i} has type {e.__class__.__name__}")
+
+            if i == 0:
+                this_upscale_schedule = e.upscale_schedule
+            else:
+                other_upscale_schedule = e.upscale_schedule
+                if np.any(this_upscale_schedule != other_upscale_schedule):
```
[the remark this answers is not in the document]
Comment (btjanaka) [reply ribs/schedulers/_bayesian_opt_scheduler.py:62]: Also, are the upscale schedules guaranteed to be numpy arrays? It seems they can be None; this comparison would fail also if either schedule is None.

File: ribs/schedulers/_bayesian_opt_scheduler.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                    if np.any(this_upscale_res != upscale_res):
+                        raise ValueError(
+                            "Emitters returned different upscale resolutions "
+                            "when they should return the same. Emitter0 "
+                            f"returned resolution {this_upscale_res}, but "
+                            f"emitter{i} returned resolution {upscale_res}")
+
+        # If the upscale resolution is not None, upscales :attr:`archive` and
+        # all emitter archives.
+        if not this_upscale_res is None:
```
Comment (btjanaka) [inline ribs/schedulers/_bayesian_opt_scheduler.py:141]: `if this_upscale_res is not None:`

File: ribs/archives/_grid_archive.py [hunk 11]
```diff
@@ -155,6 +155,55 @@ def boundaries(self):
...
+            solution_dim=self.solution_dim,
+            dims=new_dims,
+            ranges=list(zip(self.lower_bounds, self.upper_bounds)),
+            learning_rate=None,
+            threshold_min=-np.inf,
+            epsilon=self.epsilon,
+            qd_score_offset=self.qd_score_offset,
+            seed=self._seed,
+            dtype=self.dtypes,
+            extra_fields=None,
```
Comment (btjanaka) [inline ribs/archives/_grid_archive.py:197]: Why None for extra_fields?

File: ribs/archives/_grid_archive.py [hunk 12]
```diff
@@ -155,6 +155,55 @@ def boundaries(self):
...
+            is similar as the ``dims`` argument in the constructor.
+
+        Returns:
+            GridArchive: A new GridArchive with the new tessellation.
+        """
+        if not np.isclose(self.learning_rate, 1):
+            raise NotImplementedError(
+                "Cannot retessellate an archive with learning rate.")
+
+        new_archive = GridArchive(
```
Comment (btjanaka) [inline ribs/archives/_grid_archive.py:187]: I foresee this method being pretty hard to maintain. Because now whenever we change the definition of GridArchive, we have to come back here and update this method. Not quite sure how to overcome this, though. Thoughts?

File: ribs/archives/_grid_archive.py [hunk 13]
```diff
@@ -155,6 +155,55 @@ def boundaries(self):
         """
         return self._boundaries
 
+    def retessellate(self, new_dims):
```
Comment (btjanaka) [inline ribs/archives/_grid_archive.py:158]: Can you put this method in a separate PR where you also write some tests for it? Thanks!

**Review by btjanaka — CHANGES_REQUESTED** [review]
More changes. Thanks Shihan! I'll review the emitter later. I did take a look at the scheduler; it makes sense to me.

File: examples/bop_elites.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+        logdir=trial_outdir,
+        log_every=20,
+    )
+
+    # The main grid archive that interacts with BOP-Elites. We start at the
+    # lowest resolution from ``upscale_schedule``.
+    main_archive = GridArchive(
+        solution_dim=params.solution_dim,
+        dims=params.upscale_schedule[0],
+        ranges=[[-5.12, 5.12]] * 2,
```
[the remark this answers is not in the document]
Comment (szhaovas) [reply examples/bop_elites.py:142]: Are you referring to the `ranges` parameter? If I remember correctly, archive ranges define the measure space bounds and therefore depend on the measure space dimensionality. There should be another `bounds` parameter within emitters that define solution space bounds.

File: examples/bop_elites.py [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        dims=params.upscale_schedule[-1],
+        ranges=[[-5.12, 5.12]] * 2,
+        seed=params.seed,
+    )
+
+    # The main component of BOP-Elites
+    emitters = [
+        BayesianOptimizationEmitter(
+            archive=main_archive,
+            bounds=[(-10.24, 10.24)] * params.solution_dim,
```
[the remark this answers is not in the document]
Comment (szhaovas) [reply examples/bop_elites.py:159]: No, all solution space dimensions have bounds `[-10.24, 10.24]` in this example, and there are `solution_dim` such bounds.

File: ribs/schedulers/_bayesian_opt_scheduler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (szhaovas) [reply ribs/schedulers/_bayesian_opt_scheduler.py:62]: The upscale schedule is converted to numpy array when the emitter is initialized (if it is not None). 
This comparison works when either schedule is None thanks to numpy broadcasting, but it does fail when for example one schedule is `[[5, 5], [10, 10]]` but another is `[[5, 5], [10, 10], [25, 25]]`. Just fixed.

File: examples/bop_elites.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (btjanaka) [reply examples/bop_elites.py:142]: Yes I am referring to the `ranges`. For sphere, the measure space bounds (i.e., the ranges) should depend on the dimensionality of the solution space. See https://github.com/icaros-usc/pyribs/blob/master/examples/sphere.py#L744

[the remark this answers is not in the document]
Comment (btjanaka) [reply examples/bop_elites.py:142]: To be sure, we're using the same sphere as in prior work right?

**Review by btjanaka — APPROVED** [review]

File: setup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (btjanaka) [reply setup.py:42]: Following up on this, you will need to make some modifications to make the tests work -- right now the tests are failing because pymoo is not installed. You can see an example for how this is done for the pycma extra here: https://github.com/icaros-usc/pyribs/blob/master/.github/workflows/testing.yml#L52

Basically, you need to separate the tests that involve the bayesian optimization emitter into a separate test folder like `tests/emitters_pymoo`.

**Comment by btjanaka** [comment]
@szhaovas could you also add tests that, at a minimum, call the emitter and scheduler to make sure they work? For reference, you can see this emitter test https://github.com/icaros-usc/pyribs/blob/master/tests/emitters_pycma/pycma_es_test.py and the current scheduler file https://github.com/icaros-usc/pyribs/blob/master/tests/schedulers/scheduler_test.py

File: examples/bop_elites.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (szhaovas) [reply examples/bop_elites.py:142]: You are right. I forgot about the measure bounds.

File: ribs/archives/_grid_archive.py (hunk 12 above)
Comment (btjanaka) [reply ribs/archives/_grid_archive.py:194]: My concern is that it's a maintenance burden because every time we update the definition of `GridArchive`, we have to remember to modify this definition here.

File: HISTORY.md [hunk 16]
```diff
@@ -60,6 +60,9 @@ makes a couple of minor usability improvements, such as better error checking.
 - Raise error if `result_archive` and `archive` have different fields
   ({pr}`461`)
 - Warn user if resampling for bounds takes too long in ESs ({pr}`462`)
+- Implement BOP-Elites by adding BayesianOptimizationEmitter and
```
Comment (btjanaka) [inline HISTORY.md:63]: Could you move this to the top of the API section under `(Forthcoming)`? Right now, it is under the `Improvements` for `0.7.1`. We want to put it under the new release, and as API.

File: HISTORY.md [hunk 17]
```diff
@@ -60,6 +60,9 @@ makes a couple of minor usability improvements, such as better error checking.
...
+  BayesianOptimizationScheduler ({pr}`496`)
+- Add BOP-Elites demo
```
Comment (btjanaka) [inline HISTORY.md:65]: This should go under `Documentation` in the `(Forthcoming)` release as well. It should also have the PR number on it:
```
({pr}`496`)
```

File: ribs/archives/_grid_archive.py
Comment (btjanaka) [inline ribs/archives/_grid_archive.py:1]: Merge master and get rid of this change.

**Review by btjanaka — APPROVED** [review]
Fix up minor remaining comments and merge. Good job @szhaovas!

File: tests/emitters/emitter_base_test.py
[the remark this answers is not in the document]
Comment (btjanaka) [reply tests/emitters/emitter_base_test.py:1]: I would make `tests/emitters_pymoo` and `tests/schedulers_pymoo`
